Don't specialize super accessors / Fix specialization in traits

retronym opened the pull request:
(Hijacked by Adriaan.)

My hypothesis is that super call chains have to stick to methods with
the same name, because as soon as you invokevirtual another method,
your next invokevirtual will go all the way down to the dynamic type,
even though you were expecting to be in a supercall chain. Because
specialized methods are name mangled and often just delegate to the
non-specialized method, having specialized super accessors causes
problems. So, just don't do that!

We no longer emit specialized overloads of super accessors, unless
for the case tested by run/t4996.scala. Normally, we go into the
boxed world on a super call, and stay there until we hit our target.

Mixin's rebindSuper has to know about this too:

The specialized version `T$sp` of a trait `T` will have a super accessor
that has the same alias as the super accessor in trait `T`; we must
rebind super from the vantage point of the original trait `T`, not the
specialized `T$sp` (it's inserted in the base class seq late in the game
and doesn't count as a super class in the super-call scheme).

This also means rebindSuper won't rebind to specialized methods, unless
we're looking for the super of a specialized method (as that would
entail a name change during a super call chain).

Fixes scala/bug#10277
Fixes scala/bug#6265
Fixes scala/bug#8431

See 9733f56 and its commit message for an in-depth analysis

adriaanm commented:
I don't want to take one step forward and potentially two backwards, without understanding what's going on with the tests that have to be dropped. I've been investigating. Will dump some thoughts below.

adriaanm commented:
for 6265, it seems the super call is not rewritten to the specialized super method -- here's a hand-specialized version that has the bug, which blows the stack in the same way

```
trait Top[T] { def foo(x: T) = println("Top.foo " + x) }

trait M[T] extends Top[T] {
  abstract override def foo(x: T) = {
    super.foo(x) // super$foo(x)
    println("M.foo " + x)
  }
}

trait TopInt extends Top[Any] {
  def foo(x: Int) = foo_sp(x)
  def foo_sp(x: Int) = println("Top.foo " + x)
}

trait MInt extends M[Any] with TopInt {
  override def foo(x: Int) = foo_sp(x)

  abstract override def foo_sp(x: Int) = {
    super.foo(x) // should be super.foo_sp(x)
    println("M.foo " + x)
  }
}

class C extends MInt

object Test extends App {
  println((new C).foo_sp(1))
}
```

adriaanm commented:
8431 is also related to super accessors, but it seems we are overeager in emitting a forwarder for the specialized super accessor in class C

simplified test case:

```scala
trait Top[@specialized(Int) T] { def foo(x: T): Boolean = true  }
trait TopSuper[@specialized(Int) T] extends Top[T] { override def foo(x: T): Boolean = super.foo(x) }
// <superaccessor> <artifact> def super$foo(x: Object): Boolean;
// <superaccessor> <specialized> <artifact> def super$foo$mcI$sp(x: Int): Boolean = TopSuper.this.super$foo(scala.Int.box(x));

class C extends TopSuper[Int]
```

wip patch:
```diff
diff --git a/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala b/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala
index 015c415..a6569f7 100644
--- a/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala
+++ b/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala
@@ -713,18 +713,24 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
        *   def m$I(x: Int) = <body>/adapted to env {A -> Int} // om
        */
       def forwardToOverload(m: Symbol): Symbol = {
-        val specMember = enterMember(cloneInSpecializedClass(m, f => (f | OVERRIDE) & ~(DEFERRED | CASEACCESSOR)))
-        val om         = specializedOverload(sClass, m, env).setFlag(OVERRIDE)
         val original = info.get(m) match {
           case Some(NormalizedMember(tg)) => tg
           case _                          => m
         }
-        info(specMember) = Forward(om)
-        info(om)         = if (original.isDeferred) Forward(original) else Implementation(original)
-        typeEnv(om)      = env ++ typeEnv(m) // add the environment for any method tparams
-
-        newOverload(specMember, om, typeEnv(om))
-        enterMember(om)
+        if (original.hasFlag(SUPERACCESSOR)) {
+          println(s"dropping overload for $m in $sClass -- $original")
+          original
+        }
+        else {
+          val specMember = enterMember(cloneInSpecializedClass(m, f => (f | OVERRIDE) & ~(DEFERRED | CASEACCESSOR)))
+          val om = specializedOverload(sClass, m, env).setFlag(OVERRIDE)
+          info(specMember) = Forward(om)
+          info(om)         = if (original.isDeferred) Forward(original) else Implementation(original)
+          typeEnv(om)      = env ++ typeEnv(m) // add the environment for any method tparams
+
+          newOverload(specMember, om, typeEnv(om))
+          enterMember(om)
+        }
       }
 
       for (m <- normMembers ; if needsSpecialization(fullEnv, m) && satisfiable(fullEnv)) {
@@ -763,13 +769,8 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
           // was: new Forward(specMember) {
           //   override def target = m.owner.info.member(specializedName(m, env))
           // }
-        } else if (!(sClass.isTrait && m.isDeferred) && m.isMethod && !m.hasAccessorFlag) { // other concrete methods
-          // log("other concrete " + m)
+        } else if (!(sClass.isTrait && m.hasFlag(DEFERRED)) && m.isMethod && (!m.hasFlag(ACCESSOR) || m.hasFlag(LAZY)) ) { // other concrete methods
           forwardToOverload(m)
-
-        } else if (!(sClass.isTrait && m.isDeferred) && m.isMethod && m.hasFlag(LAZY)) {
-          forwardToOverload(m)
-
         } else if (m.isValue && !m.isMethod) { // concrete value definition
           def mkAccessor(field: Symbol, name: Name) = {
             val newFlags = (SPECIALIZED | m.getterIn(clazz).flags) & ~(LOCAL | CASEACCESSOR | PARAMACCESSOR)
```

adriaanm commented:
(my horrible wip commits are on the same branch on my fork -- mainly, why do we even specialize super accessors? can we fix the invokespecial behavior by using the right overload, as in the hand-specialized version above?)

adriaanm committed:
Don't specialize super accessors

My hypothesis is that super call chains have to stick to methods with
the same name, because as soon as you invokevirtual another method,
your next invokevirtual will go all the way down to the dynamic type,
even though you were expecting to be in a supercall chain. Because
specialized methods are name mangled and often just delegate to the
non-specialized method, having specialized super accessors causes
problems. So, just don't do that!

We no longer emit specialized overloads of super accessors, unless
for the case tested by run/t4996.scala. Normally, we go into the
boxed world on a super call, and stay there until we hit our target.

Mixin's rebindSuper has to know about this too:

The specialized version `T$sp` of a trait `T` will have a super accessor
that has the same alias as the super accessor in trait `T`; we must
rebind super from the vantage point of the original trait `T`, not the
specialized `T$sp` (it's inserted in the base class seq late in the game
and doesn't count as a super class in the super-call scheme).

This also means rebindSuper won't rebind to specialized methods, unless
we're looking for the super of a specialized method (as that would
entail a name change during a super call chain).

adriaanm commented:
I made some progress on my branch, but 8431 is still failing. I'm still worried that we're too close to RC to be messing with specialization, especially since this already regressed in 2.12. I think we should target 2.12.9 instead

adriaanm commented:
I don't know how to fix this further. My hunch is that it's related to invokespecial and jumping between method names (the `foo$sp` to `foo` delegation) during a super-call chain. When we "exit" the super call chain and call `foo`, we actually go back to invokevirtual style dispatch, and end up calling the method in C instead of the delegatee of the super call in Top.

adriaanm commented:
the loop in 8431
```
at C.foo$mcI$sp(t8431.scala:9)
// invokeinterface foo$mcI$sp:(I)Z
at Mid$mcI$sp.foo(t8431.scala:6)
// invokespecial foo:(I)Z
at Mid$mcI$sp.foo$(t8431.scala:6)
// invokestatic Mid$mcI$sp.foo$:(LMid$mcI$sp;I)Z
at C.foo(t8431.scala:9)
// invokevirtual foo:(I)Z
at C.foo(t8431.scala:9)
// invokeinterface foo:(Ljava/lang/Object;)Z
at Top.foo$mcI$sp(t8431.scala:2)
//invokespecial foo$mcI$sp:(I)Z
at Top.foo$mcI$sp$(t8431.scala:2)
// invokestatic Top.foo$mcI$sp$:(LTop;I)Z
at C.super$foo$mcI$sp(t8431.scala:9)
// invokeinterface super$foo$mcI$sp:(I)Z
at Mid$mcI$sp.foo$mcI$sp(t8431.scala:6)
// invokespecial foo$mcI$sp:(I)Z
at Mid$mcI$sp.foo$mcI$sp$(t8431.scala:6)
// invokestatic  Mid$mcI$sp.foo$mcI$sp$:(LMid$mcI$sp;I)Z
at C.foo$mcI$sp(t8431.scala:9)
```

adriaanm commented:
That made a 💡 turn on. I pushed a new approach -- feeling reasonably good about it.

adriaanm commented:
/synch

adriaanm commented:
/rebuild

retronym commented:
Thanks for pulling the thread. I'll review today. In general, changes to specialization work best in a major release if we want to support separate compilation of parts of a specialized hierarchy.

retronym reviewed (commented):
LGTM (I can't approve my own PR)
